File: app/src/main/java/com/ndmsystems/coala/CoAPServer.java
package com.ndmsystems.coala;

import com.ndmsystems.coala.message.CoAPRequestMethod;
import com.ndmsystems.coala.resource_discovery.ResourceDiscoveryResult;

import java.util.List;

import io.reactivex.Single;

public interface CoAPServer {
    interface OnResourcesDiscovered {
        void onResourcesDiscovered(List<ResourceDiscoveryResult> resourceDiscoveryResults);
    }

    CoAPObservableResource getObservableResource(String path);

    void addObservableResource(String path, CoAPResource.CoAPResourceHandler handler);

    void addResource(String path, CoAPRequestMethod method, CoAPResource.CoAPResourceHandler handler);

    void removeResource(String path, CoAPRequestMethod method);

    Single<List<ResourceDiscoveryResult>> runResourceDiscovery();
}

File: app/src/main/java/com/ndmsystems/coala/CoAPServer.kt
package com.ndmsystems.coala

import com.ndmsystems.coala.CoAPResource.CoAPResourceHandler
import com.ndmsystems.coala.message.CoAPRequestMethod
import com.ndmsystems.coala.resource_discovery.ResourceDiscoveryResult

interface CoAPServer {

    fun getObservableResource(path: String): CoAPObservableResource?
    fun addObservableResource(path: String, handler: CoAPResourceHandler)
    fun addResource(path: String, method: CoAPRequestMethod, handler: CoAPResourceHandler)
    fun removeResource(path: String, method: CoAPRequestMethod)
    suspend fun runResourceDiscovery(): List<ResourceDiscoveryResult>
}